app/scripts/controllers/user-storage/encryption.ts

```diff
@@ -0,0 +1,117 @@
+import sjcl from 'sjcl';
```

Comment:
I think we might need a wider discussion on this.

## Context

All MetaMask Clients (mobile, extension, portfolio) that want to use UserStorage must all use the same encryption/decryption mechanism.

Any data you store must be encrypted, and any data you retrieve must be decrypted.

This is what was approved on Portfolio - basic symmetric encryption using the storage key as the password.

## Discussion

Can we also use this for extension? What are the risks? If there are other encryption mechanisms that we need, we will need it to be compatible for all metamask clients and also need to port Portfolio to use the same encryption mechanism.

Reply:
Is the storage key a symmetric key?

Also, we should rather use a KDF instead of a PBKDF if we are already working with cryptographic material.

Reply:
Yep the storage key is symmetric. And yep flexible to change to KDF.
TBH if there is a need to change to asymmetric encryption/decryption I think we can - I'll note this.

Reply:
I don't see a usecase for assymetric cryptography here. Care to elaborate?

Reply:
I agree - I don't see a need for asymmetric encryption - was open to it if it is recommended.

Reply:
Thanks all! Closing this discussion.

As discussed we have a Threat Modelling Document [here](https://github.com/MetaMask/threatmodels/pull/8)
Also tickets [here](https://github.com/MetaMask/threatmodels/pull/8)